astropy/coordinates/transformations.py

```diff
@@ -744,8 +747,111 @@ def __call__(self, fromcoord, toframe):
                 'should have been of type {1}'.format(res, self.tosys))
         return res
 
+class BaseAffineTransform(CoordinateTransform):
+    """
+    Base class for common functionality between the ``AffineTransform``-type
+    subclasses.
+    """
+
+    def _apply_transform(self, fromcoord, matrix, offset):
+        from .representation import (CartesianRepresentation,
+                                     UnitSphericalRepresentation,
+                                     CartesianDifferential)
 
-class StaticMatrixTransform(CoordinateTransform):
+        rep = fromcoord.data.to_cartesian_with_differentials()
+
+        # Only do transform is matrix is specified. This is for speed in
+        # transformations that only specify an offset (e.g., LSR)
+        if matrix is not None:
+            # Note: this applies to both representation and differentials
+            rep = rep.transform(matrix)
+
+        # TODO: if we decide to allow arithmetic between representations that
+        # contain differentials, this can be tidied up
+        if offset is not None:
+            newrep = (rep.without_differentials() +
+                      offset.without_differentials())
+        else:
+            newrep = rep.without_differentials()
+
+        # We need a velocity (time derivative) and, for now, are strict: the
+        # representation can only contain a velocity differential and no others
+        if 's' in rep.differentials and len(rep.differentials) == 1:
+            veldiff = rep.differentials['s'] # already in Cartesian form
+
+            if 's' in offset.differentials:
+                veldiff = veldiff + offset.differentials['s']
+
+            newrep = newrep.with_differentials(
+                veldiff.represent_as(rep.differentials['s'].__class__, newrep))
+
+        elif len(rep.differentials) > 1:
+            # We should never get here because the frame initializer sholdn't
+            # allow more differentials, but this just adds protection for
+            # subclasses that somehow skip the checks
+            raise ValueError("Representation passed to AffineTransform contains"
+                             " multiple associated differentials. Only a single"
+                             " differential with velocity units is presently"
+                             " supported (differentials: {0})."
+                             .format(str(rep.differentials)))
+
+        if isinstance(fromcoord.data, UnitSphericalRepresentation):
+            # need to special-case this because otherwise the new class will
```

Comment:
I'm actually confused - why would there be confusion?
More generally, should we always transform back to whatever the input representation and differential were? Or is it sometimes not wanted? ... looking further down, I guess the idea is that there's no need to convert to, say, Spherical, until someone actually wants something with the coordinates. So, this is just so one doesn't have to check the unit on the cartesian representation to know whether it is unitspherical. I almost wonder if we should have a little "marker" on cartesian... Anyway, I think I understand it now, so if the above is indeed correct, no action needed (except if you think you can clarify the comment further).